File: _permutations.ts
function permute(nums: number[]): number[][] {

    let res =[]
    let used = []

    let backtrack = (used,cur)=>{
        
        if(cur.length == nums.length){
            res.push(Array.from(cur))
            // console.log('res',res)
            return
        }

        for(let j=0; j<nums.length; j++){
            
            if(used.includes(nums[j])){
                continue
            }

            used.push(nums[j])
            cur.push(nums[j])
            backtrack(used,cur)
            used.pop()
            cur.pop()
        }
    }

    backtrack([],[])
    return res

};

File: Python/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:

        res = []
        n = len(nums)

        def back(i,cur,visit):
            nonlocal n
            if i>=n:
                res.append(cur.copy())
                return
            
            for num in nums:
                if( num not in visit):
                    cur.append(num)
                    visit.append(num)

                    back(i+1,cur,visit)

                    cur.pop()
                    visit.pop()



        for i in nums:
            visit= []
            cur = []
            cur.append(i)
            visit.append(i)
            back(1,cur,visit)
        
        return res
        

            
